### ISSUE TITLE
deadlock producing events from timer

### ISSUE THREAD

**Issue by balazsracz** [bakerstu/openmrn#9]
I wrote a small piece of code that polls a GPIO pin from a recurring timer, and does nmranet_event_produce if it sees some changes.

This code deadlocked upon initialization. The deadlock occurred on nodeMutex.

The mutex holder was thread.main with the following stacktrace:

#0  sleep (seconds=seconds@entry=1) at ../../..//src/os/os.c:732
#1  0x00001858 in upstream_alias_setup (node_id=11936128518282651045, can_if=0x8002ccca)
    at ../../..//src/if/nmranet_can_if.c:436
#2  can_write (nmranet_if=0x8002ccca, mti=<optimized out>, src=<optimized out>, dst=..., data=0x1000563c)
    at ../../..//src/if/nmranet_can_if.c:496
#3  0x00004262 in nmranet_node_write (node=node@entry=0x10004ff0, mti=mti@entry=368, dst=..., data=data@entry=0x1000563c)
    at ../../..//src/core/nmranet_node.c:555
#4  0x0000432c in verify_node_id_number (node=node@entry=0x10004ff0) at ../../..//src/core/nmranet_node.c:293
#5  0x00004788 in nmranet_node_initialized (node=0x10004ff0) at ../../..//src/core/nmranet_node.c:181
#6  0x0000036c in appl_main (argc=argc@entry=1, argv=argv@entry=0x200809f8) at ../../main.cxx:300
#7  0x00004832 in main_thread (arg=<optimized out>) at ../../..//src/os/os.c:837

And the timer Svc deadlocked with the following stacktrace:
#0  0x00005436 in xQueueGenericReceive (xQueue=0x8002ccca, xQueue@entry=0x10004fa0, pvBuffer=0xa5a5a5a5, 
    pvBuffer@entry=0x0, xTicksToWait=xTicksToWait@entry=4294967295, xJustPeeking=-1515870811, xJustPeeking@entry=0)
    at /home/bracz/FreeRTOS/Source/queue.c:1146
#1  0x0000548a in xQueueTakeMutexRecursive (xMutex=0x10004fa0, xBlockTime=xBlockTime@entry=4294967295)
    at /home/bracz/FreeRTOS/Source/queue.c:508
#2  0x000036ec in os_mutex_lock (mutex=0x1000014c) at ../../..//src/os/os.h:383
#3  os_mutex_lock (mutex=0x1000014c) at ../../..//src/os/os.h:364
#4  0x00003a54 in nmranet_event_produce (node=0x10004ff0, event=<optimized out>, state=0)
    at ../../..//src/core/nmranet_event.c:294
#5  0x00000502 in MemoryBitProducer<unsigned long volatile>::PerformUpdate (this=0x20080980)
    at ../../src/common_event_handlers.hxx:145
#6  0x000005cc in TimerUpdater::callback (object=<optimized out>) at ../../src/timer_updater.hxx:27
#7  0x000048a4 in timer_callback (timer=0x100055b0) at ../../..//src/os/os.c:164
#8  0x00006240 in prvProcessExpiredTimer (xTimeNow=200, xNextExpireTime=200) at /home/bracz/FreeRTOS/Source/timers.c:347
#9  prvProcessTimerOrBlockTask (xListWasEmpty=<optimized out>, xNextExpireTime=200)
    at /home/bracz/FreeRTOS/Source/timers.c:394
#10 prvTimerTask (pvParameters=<optimized out>) at /home/bracz/FreeRTOS/Source/timers.c:367

after a few seconds my watchdog kills the process.

I don't understand the specific details of the alias setup routine, but I guess that the timer svc being blocked on a mutex probably means that the nmranet alias setup process will never terminate and thus will never release the nodeMutex.

**Comment by bakerstu** [comment]
The fundamental architecture has changed.  This issue is no longer relevant.
